core/embed/rust/src/ui/model_tt/component/progress.rs

```diff
@@ -0,0 +1,136 @@
+use core::mem;
+
+use crate::ui::{
+    component::{
+        base::ComponentExt,
+        paginated::Paginate,
+        text::paragraphs::{Paragraph, Paragraphs},
+        Child, Component, Event, EventCtx, Label, Never, Pad,
+    },
+    display::{self, Font},
+    geometry::{Insets, Rect},
+    model_tt::constant,
+    util::animation_disabled,
+};
+
+use super::theme;
+
+pub struct Progress<T, F>
+where
+    F: Fn() -> T,
+{
+    title: Child<Label<T>>,
+    value: u16,
+    loader_y_offset: i16,
+    indeterminate: bool,
+    description: Child<Paragraphs<Paragraph<T>>>,
+    description_pad: Pad,
+    description_func: F,
+}
+
+impl<T, F> Progress<T, F>
+where
+    T: AsRef<str>,
+    F: Fn() -> T,
+{
+    const AREA: Rect = constant::screen().inset(theme::borders());
+
+    pub fn new(title: T, indeterminate: bool, description_func: F) -> Self {
+        Self {
+            title: Label::centered(title, theme::label_progress()).into_child(),
+            value: 0,
+            loader_y_offset: 0,
+            indeterminate,
+            description: Paragraphs::new(
+                Paragraph::new(&theme::TEXT_NORMAL, description_func()).centered(),
+            )
+            .into_child(),
+            description_pad: Pad::with_background(theme::BG),
+            description_func,
+        }
+    }
+}
+
+impl<T, F> Component for Progress<T, F>
+where
+    T: AsRef<str>,
+    F: Fn() -> T,
+{
+    type Msg = Never;
+
+    fn place(&mut self, _bounds: Rect) -> Rect {
+        let description_lines = 1 + self
+            .description
+            .inner()
+            .inner()
+            .content()
+            .as_ref()
+            .chars()
+            .filter(|c| *c == '\n')
+            .count() as i16;
+        let (title, rest) = Self::AREA.split_top(self.title.inner().size().y);
+        let (loader, description) =
+            rest.split_bottom(Font::NORMAL.line_height() * description_lines);
+        let loader = loader.inset(Insets::top(theme::CONTENT_BORDER));
+        self.title.place(title);
+        self.loader_y_offset = loader.center().y - constant::screen().center().y;
+        self.description.place(description);
+        self.description_pad.place(description);
+        Self::AREA
+    }
+
+    fn event(&mut self, ctx: &mut EventCtx, event: Event) -> Option<Self::Msg> {
+        if let Event::Progress(new_value) = event {
+            if mem::replace(&mut self.value, new_value) != new_value {
+                if !animation_disabled() {
+                    ctx.request_paint();
+                }
+                let new_description = (self.description_func)();
```

Comment:
Do we currently have some use-case of `description_func` changing its return value between calls, or it is just a nice-to-have possibility for the future?

Bare function without arguments has not an easy job of tracking the progress easily - to return some appropriate description each time (it needs to store context in its global environment). What if we supplied the `new_value` into it, so the function could have some logic inside to response to changing (increasing) progress.

Simple example:
```python
def description(value: int) -> str:
    if value < 100:
        return "Beginning message"
    elif value < 900:
        return "Middle message"
    else:
        return "Finishing message"
```

Reply:
PIN unlock progress changes the description to display number of seconds remaining, also Monero signing uses this.

The logical place for description updates would be in `Event::Progress(...)` but we currently don't have a way to send a python string along with an event and I didn't want to add it just for this. Using a closure for `description_func` still allows the caller (RustProgress) to provide the description when it changes.

Reply:
Right, I did not notice that client can be changing the description at any time in `RustProgress::report`, so  `(lambda: self.description)()` can return different results at any time.